fix handling of commas in X509 Strings

RonEld opened the pull request:
According to RFC 1779, commas in strings should be either in quotes,
or preceded by escape character.
Add escape charatcer to the string representation of a distinguished name, in case there is comma
Raised in #769 by cwtios

* Needs backporting to 2.1 and  2.7
* added test case with quotes in CN
* added some tests for invalid locations of commas
* added test case for comma in subject \ issuer name, and its string representation

**Note: This PR handles commas only, and no other special characters defined in RFC 1779, and currently not supported**

andresag01 commented on library/x509_create.c:
```diff
@@ -143,6 +149,9 @@ int mbedtls_x509_string_to_names( mbedtls_asn1_named_data **head, const char *na
 
         if( !in_tag && s != c + 1 )
         {
+            if( *c == '\"' )
```
I do not think this works well in some cases. If I understand correctly, you are trying to match pairs of `"` characters so that `,` are note interpreted as separators. However, this does not consider a distinguished name of the form:
```
C=NL,O=Of\"fspark\, Inc., OU=PolarSSL
```
In this case the`"` should be interpreted as part of the organisation instead of the start of a quoted string.

RonEld replied:
correct, but at the moment, we only escape the comma.

andresag01 commented on tests/suites/test_suite_x509write.data:
```diff
@@ -59,19 +59,22 @@ depends_on:MBEDTLS_SHA1_C:MBEDTLS_RSA_C:MBEDTLS_PKCS1_V15:MBEDTLS_DES_C:MBEDTLS_
 X509 String to Names #2
+mbedtls_x509_string_to_names:"C=NL,O="Offspark, Inc.", OU=PolarSSL":"C=NL, O="Offspark, Inc.", OU=PolarSSL":0
+
+X509 String to Names #3
```
Perhaps instead of changing all the test numbers you could simply add new tests with higher numbers that currently do not exist? e.g. 7, 8, 9, etc. Reducing the number of changes makes the important code changes easier to see.

RonEld replied:
ok, my idea was to add the test case in the context of the previous test. I will change it according to your comment

andresag01 commented on tests/suites/test_suite_x509write.data:
```diff
@@ -59,19 +59,22 @@ depends_on:MBEDTLS_SHA1_C:MBEDTLS_RSA_C:MBEDTLS_PKCS1_V15:MBEDTLS_DES_C:MBEDTLS_
+mbedtls_x509_string_to_names:"C=NL,O=Offspark\, Inc., OU=PolarSSL":"C=NL, O=Offspark\, Inc., OU=PolarSSL":0
 
 X509 String to Names #2
+mbedtls_x509_string_to_names:"C=NL,O="Offspark, Inc.", OU=PolarSSL":"C=NL, O="Offspark, Inc.", OU=PolarSSL":0
```
I think these two tests are a good starting point, but perhaps we could add some other interesting test cases? For example escaping `"`, or using other separators.

RonEld replied:
It depends whether we decide to support escaped `"` in the x509 certificate, according to previous comment

RonEld commented on library/x509_create.c:
```diff
@@ -95,6 +95,7 @@ int mbedtls_x509_string_to_names( mbedtls_asn1_named_data **head, const char *na
     int in_tag = 1;
     char data[MBEDTLS_X509_MAX_DN_NAME_SIZE];
     char *d = data;
+    int in_quot=0;
```
ok

RonEld commented on library/x509_create.c:
```diff
@@ -116,16 +117,21 @@ int mbedtls_x509_string_to_names( mbedtls_asn1_named_data **head, const char *na
+                ret = MBEDTLS_ERR_X509_INVALID_NAME;
+                goto exit;
+            }
+            *(d++) = *c;
```
both write and read of x509 certificate get into this function, so the parsing will be done correct

RonEld commented on library/x509_create.c:
```diff
@@ -116,16 +117,21 @@ int mbedtls_x509_string_to_names( mbedtls_asn1_named_data **head, const char *na
 
         if( !in_tag && *c == '\\' && c != end )
```
agreed on first coment, on second comment, we rely on `c == end` so we would store the named data, in line 134

AndrzejKurek replied:
!in_tag could be extracted as a separate condition above, and these three separate checks for it would be unnecessary.

RonEld commented:
@andresag01 I modified according to your comments.
We still need to decide whether we support other characters that need to be escaped according to RFC1779

yanesca commented:
I am sorry, but I don't think that the new behaviour is correct. The certificate will be issued to *Offspark\, Inc.* instead of *Offspark, Inc.*.

We should keep in mind that RFC1779 describes the string representation of Distinguished Names and not the DNs themselves! Escaping the commas or quoting them is just a way to represent them in a string that can unambiguously parsed. The attribute encoded in ASN1 in the certificate is without quotes or escape characters. With this fix the quotes and escaping backslashes spill into the ASN1 encoding as well.

Therefore I think that the functionality of `mbedtls_x509_string_to_names()` is correct in handling escaped commas and we need to leave it unchanged.

The current implementation of `mbedtls_x509_dn_gets()` in turn is incorrect and updating it to escape the names properly would fix the issue at hand.

yanesca commented:
@andresag01 You are right, the code implements only a small subset of RFC1779. This might not be optimal if we are looking at Mbed TLS as a whole, however I don't think that this is a problem in the case of this specific PR.

Also, I share your feelings about the second point you raised, the goal of this PR is to fix a bug and not to introduce a new feature. We shouldn't be mixing the two, unless the new feature is relatively small and essential to the fix.

sbutcher-arm commented:
Moving the review from @yanesca to @AndrzejKurek, to ease the review burden on @yanesca.

RonEld commented:
I disagree with @yanesca statement on the string representation of the Distinguished names.
[RFC1779 section 2.2](https://tools.ietf.org/html/rfc1779#section-2.2) states:
> The final examples show both methods quoting of a comma in an
   Organisation name:
   CN=L. Eagle, O="Sue, Grabbit and Runn", C=GB
   CN=L. Eagle, O=Sue\, Grabbit and Runn, C=GB

So, the CN should be either `CN="Offspark, inc."` or `CN=Offspark\, Inc.` according to the RFC

yanesca commented:
I am not debating what the standard says, I am just saying that RFC1779 describes how a string representation of the DN should look like and not what a DN is. The definition of what a DN is is defined in X500.

The function `mbedtls_x509_string_to_names()` takes a string representation defined in RFC1778 and converts it to an internal representation that strongly resembles to what is defined in X500. Then later this internal representation gets encoded in ASN1.

Because of this, if we include the comma in the internal representation, then it gets included in the ASN1 representation. Therefore if we have an entity that is called `Offspark, inc.` , we issue a certificate to an entity called `"Offspark, inc."` or `Offspark\, Inc.`.

RonEld commented:
OK, so maybe I misunderstood you. Nonetheless, we should be able to parse both representations

yanesca commented:
Yes, it can be a useful feature and now that it is there I am not suggesting to remove it.

(I wasn't requesting it before either, I just stated my general preference of having bug fixes and new features in separate PRs. And just to be clear: I wasn't and I am not suggesting to split it up into two PRs either.)

AndrzejKurek commented on library/x509_create.c:
```diff
@@ -95,6 +95,7 @@ int mbedtls_x509_string_to_names( mbedtls_asn1_named_data **head, const char *na
     int in_tag = 1;
     char data[MBEDTLS_X509_MAX_DN_NAME_SIZE];
     char *d = data;
+    int in_quot = 0;
```
Why not change it to a boolean if we're only doing logic operations on it?

yanesca replied:
Do we have defined a boolean type somewhere?

sbutcher-arm replied:
We don't have a boolean type, and an `int` is usually used.

yanesca commented:
As far as I can tell this PR hasn't been changed since my comments and I still think that the current implementation of `mbedtls_x509_string_to_names()` is correct and `mbedtls_x509_dn_gets()` is the one we need to fix.

sbutcher-arm commented on tests/suites/test_suite_x509write.data:
```diff
@@ -59,7 +59,7 @@ depends_on:MBEDTLS_SHA1_C:MBEDTLS_RSA_C:MBEDTLS_PKCS1_V15:MBEDTLS_DES_C:MBEDTLS_
 
 X509 String to Names #1
-mbedtls_x509_string_to_names:"C=NL,O=Offspark\, Inc., OU=PolarSSL":"C=NL, O=Offspark, Inc., OU=PolarSSL":0
+mbedtls_x509_string_to_names:"C=NL,O=Offspark\, Inc., OU=PolarSSL":"C=NL, O=Offspark\, Inc., OU=PolarSSL":0
```
I think we should have an additional test case of an unescaped comma, and an escaped comma followed by an unescaped comma.

RonEld commented:
@yanesca OK, I agree.
However, when I try the following:
~~~
./cert_write selfsign=1 issuer_name="C=UK O=test, inc. OU=testsuite CN=localhost" issuer_key=../../tests/data_files/ec_prv.sec1.der not_before=20130101000000 not_after=20151231235959 is_ca=0 max_pathlen=0 output_file=cert_example_commas.crt
~~~
The operation failed, regardless of `mbedtls_x509_dn_gets()`, because the program calls:
~~~
mbedtls_x509write_crt_set_subject_name( &crt, opt.subject_name )
~~~
ehich uses the input as the subject name \ issuer name, and within it calls `mbedtls_x509_string_to_names()`.
Do you think it's an issue in the program? (not calling mbedtls_x509_dn_gets )? Do you think it's a bug in the `mbedtls_x509write_crt_set_subject_name` \ `mbedtls_x509write_crt_set_issuer_name` ?
I think it's a valid parameter, because it's within a quote

yanesca commented:
@RonEld `mbedtls_x509_dn_gets()` is not called, because you are making a self-signed certificate. This function is used when retrieving a name in a printable representation from the internal representation.

Also the quotes are stripped by the shell, escape them or use an additional pair of single quotes to protect them:
```sh
./cert_write selfsign=1 issuer_name='"C=UK O=test, inc. OU=testsuite CN=localhost"' issuer_key=../../tests/data_files/ec_prv.sec1.der not_before=20130101000000 not_after=20151231235959 is_ca=0 max_pathlen=0 output_file=cert_example_commas.crt
```

RonEld commented:
@yanesca I made a mistae in my command.
It should be:
~~~
./cert_write selfsign=1 issuer_name="C=UK O=\"test,inc.\" OU=testsuite CN=localhost" issuer_key=../../tests/data_files/ec_prv.sec1.der not_before=20130101000000 not_after=20151231235959 is_ca=0 max_pathlen=0 output_file=cert_example_commas2.crt
~~~
I did a self signed certificate for convenience. But the behavior would be the same for a regular certificate

yanesca commented:
Still not correct, try:
```
./cert_write selfsign=1 issuer_name="C=UK, O=\"test,inc.\", OU=testsuite, CN=localhost" issuer_key=../../tests/data_files/ec_prv.sec1.der not_before=20130101000000 not_after=20151231235959 is_ca=0 max_pathlen=0 output_file=cert_example_commas2.crt
```

RonEld commented:
@yanesca Thanks, it was a typo on my side. it works now ( at least after I fixed it in https://github.com/ARMmbed/mbedtls/pull/784/commits/a95fabcbbf3f282eb3bbed8ac61598669e77b5d5)

RonEld commented:
@andresag01 @AndrzejKurek @sbutcher-arm I have changed my fix according to comments from @yanesca .
Please review

RonEld commented:
I have also added a couple of tests, as commented in https://github.com/ARMmbed/mbedtls/pull/784#discussion_r196483882

andresag01 commented on ChangeLog:
```diff
@@ -29,6 +29,11 @@ Bugfix
+   * According to RFC 1779, commas in strings should be either in quotes,
+     or preceded by escape character. Change mbedtls_x509_dn_gets() to add a '\'
+     character before a comma, in case it is part of the tag value. In addition,
+     modify mbedtls_x509_string_to_names() to ignore commas as seperators
```
Typo: seperators -> separators

andresag01 commented on library/x509.c:
```diff
@@ -780,10 +781,16 @@ int mbedtls_x509_dn_gets( char *buf, size_t size, const mbedtls_x509_name *dn )
+            {
+                s[i + j] = '\\';
+                j++;
+                s[i + j] = c;
```
I think there is a change this might overflow the buffer s. Note that in the previous version of the code at most 1 character was written per loop iteration, so the following check at the begining of each loop iteration was sufficient:
```
             if( i >= sizeof( s ) - 1 )
                 break;
```
I think that with the new version of the code, if that check passes, but there is only 1 byte of space left, then you will write 2 byte if the next character is a ,

RonEld replied:
Good catch!
It should be:
~~~
if( i + j >= sizeof( s ) - 1 )
    break;
~~~

andresag01 replied:
Thanks for looking into it, but IMHO the new condition is not sufficient. For example, consider the case where at the beginning of an iteration in line 779 the condition `i + j >= sizeof( s ) - 1` is true but the condition `i + j + 1 >= sizeof( s ) - 1` is not. The check in line 779 will succeed. Then, if the character at `name->val.p[i]` is a `,` the program will eventually reach line 787 and the following two memory locations will be written:
```
s[i + j] = '\\';
s[i + j + 1] = c; 
```
Finally, in the following iteration the condition in line 779 will not be true and the location s[i + j] will be written with the value '\0', but at this point both i and j have been incremented by 1 so in fact the program is writing location s[i + j + 2] with regards to the values for i and j from the previous iteration causing a memory overwrite of 1 byte.

RonEld replied:
Good point! I'll add additional checks for overflow

andresag01 commented on library/x509_create.c:
```diff
@@ -112,6 +113,19 @@ int mbedtls_x509_string_to_names( mbedtls_asn1_named_data **head, const char *na
+        {
+            c++;
+            s = c + 1;
+            in_quot = 0;
```
Is it possible for this code to overread from the buffer pointed to by s and c? That is, the check at the begining of the loop (`c <= end`) does not guarantee anything other than "we can read the character at c". For example in line 116 the new code reads the location at c+1 without checking that `c+1 <= end`, but I am not familiar enough with the code to know that there exists a path through the code that causes such errors. What do you think?

I think that tests like this are on the border of fuzzing, but perhaps we could create a couple of tests that check corner cases? For example, an unterminated quote or a name that is empty after the equal and is at the end of the string.

RonEld replied:
I see your point. I'll see how I can fix it.
In addition, I think the whole code should probably change from `if ( c != end )` to `if( c <= end )`.
What do you think?
I'll add some more corner case checks as well

RonEld commented:
@andresag01 I addressed your comments.
I also added a couple of corner cases as you requested, however I don't think they are part of thescope of this PR.
I think the function `mbedtls_x509_string_to_names()` should be refactored for such corner cases, and supported characters. I created a task in #1865

andresag01 commented on library/x509.c:
```diff
@@ -753,6 +753,7 @@ int mbedtls_x509_dn_gets( char *buf, size_t size, const mbedtls_x509_name *dn )
 
     while( name != NULL )
     {
+        j = 0;
```
Perhaps we could move this initialisation closer to the for loop where the variable is used? used?

RonEld replied:
Perhaps, personally, I prefer to  have variables initialized at the first place that is possible, but I don't have strong feelings on that

RonEld commented:
@andresag01 Thank you for your comments!
I believe I have addressed them.

sbutcher-arm commented:
@andresag01 / @AndrzejKurek - Could you please re-review?

andresag01 commented on ChangeLog:
```diff
@@ -29,6 +29,11 @@ Bugfix
+     or preceded by escape character. Change mbedtls_x509_dn_gets() to add a '\'
+     character before a comma, in case it is part of the tag value. In addition,
+     modify mbedtls_x509_string_to_names() to ignore commas as separators
+     in case they are within a quoted string. Raised in #769 by cwtios.
```
Minor: I suggest rewording the last sentence slightly:
>In addition, modify mbedtls_x509_string_to_names() to not interpret commas as separators when they are within a quoted string.

RonEld commented:
@andresag01 I updated the ChangLog according to your suggestion.
I think that refactoring the function merits a different PR( Perhaps as a fix for #1865 )

RonEld commented:
I created backports to versions 2.1 and 2.7 in #2048 and #2049 respectively

hanno-arm commented on ChangeLog:
```diff
@@ -29,6 +29,11 @@ Bugfix
      by Brendan Shanks. Part of a fix for #992.
    * Fix compilation error when MBEDTLS_ARC4_C is disabled and
      MBEDTLS_CIPHER_NULL_CIPHER is enabled. Found by TrinityTonic in #1719.
+   * According to RFC 1779, commas in strings should be either in quotes,
```
Style: Use imperative style and summarize the change in one sentence. Suggestion: `Adapt X.509 DN handling to conform to RFC 1779. Specifically, ...` or `Adapt X.509 DN handling to conform to RFC 1779: ...`

hanno-arm commented on library/x509.c:
```diff
@@ -773,17 +773,26 @@ int mbedtls_x509_dn_gets( char *buf, size_t size, const mbedtls_x509_name *dn )
         for( i = 0; i < name->val.len; i++ )
         {
-            if( i >= sizeof( s ) - 1 )
+            if( i + j >= sizeof( s ) - 1 )
```
I second @AndrzejKurek's comment on readability: Could we replace `i + j` by an `offset` variable which points to the current write-position into the target `s` buffer, and which is increased with every write? I think this make the code much easier to read.

RonEld commented:
@hanno-arm Thank you for your review
I will remake the fix.
As for your third comment:
"The PR doesn't get more complex by adding support for quoting other special characters apart from ,; in fact, I think it would get easier to read if we had an array special_symbols or similar against which characters would be checked."
It's not a matter of complexity. It's a matter that fixing `,` is a bug fix, while supporting other special characters is an enhancement. Having an array of `spacial_symbols` was my idea as well, for resolving #1865

RonEld commented:
@hanno-arm I addressed what I believe is the bugfix ( comments 1 and 2 )

IMHO, comment 3 is an enhancement, and merits a different PR.
Please review

hanno-arm commented on library/x509_create.c:
```diff
@@ -86,62 +92,141 @@ static const char *x509_at_oid_from_name( const char *name, size_t name_len )
 
-        if( !in_tag && s != c + 1 )
+        if( ( dn_cur_state != dn_in_tag_no_quot ) &&
+            ( s != c + 1 ) )
```
To my understanding, this is supposed to copy any character within a DN value into the temporary `data` buffer. However, it doesn't work for e.g. for a DN name `A=""`, in which case the correct behaviour would be to parse an empty quoted string (and error out); instead, the code would read the second `"` into the `data` buffer.

I suggest to move the copying code into the sections handling the states `dn_in_quoted_val` and `dn_in_unquoted_value`, once it is known that the next character does not lead to a state change. This way, the logic is very simple: For any state, check if the current character is state-changing; if yes, change the state; if not, potentially unescape it and copy it to the target buffer (a no-op for the `s` buffer, and an actual copy for the `d` buffer).

RonEld replied:
I have added a test case for this(not committed yet), and we return a failure.

hanno-arm replied:
Yes, but the internal treatment is incorrect nonetheless; we should never copy a delimiting quote into the DN value buffer.

RonEld replied:
Yes, I see your point. I'll fix this

hanno-arm replied:
I find this still hard to read. Suggestion: Add a local variable `copy_cur` or similar which is set by the state handling for `dn_in_quoted_value` and `dn_in_unquoted_value` if they found a character that should be copied to the value buffer `data`, and use that here.

RonEld replied:
I don't quite understand what you mean with your last comment

RonEld commented on library/x509_create.c:
```diff
@@ -39,6 +39,12 @@ typedef struct {
 } x509_attr_descriptor_t;
 
+typedef enum {
+    dn_no_tag_no_quot = 0,
```
I agree. I am not keen on these names as well. Your suggestion is better

RonEld commented on library/x509_create.c:
```diff
@@ -86,62 +92,141 @@ static const char *x509_at_oid_from_name( const char *name, size_t name_len )
+                {
+                    c++;
+                    s = c + 1;
+                    if( *c != ',' )
```
I'll skip whitespaces

RonEld replied:
Reading the RFC again, only one optional space is allowed. I will check only one space

hanno-arm replied:
@RonEld I disagree, the RFC contains the following grammar snippet:
```
<spaced-separator> ::= <optional-space>
                   <separator>
                   <optional-space>

   <separator> ::=  "," | ";"

   <optional-space> ::= ( <CR> ) *( " " )
```

hanno-arm replied:
See [RFC 822, Sect. 2.4](https://tools.ietf.org/html/rfc822#section-2).

RonEld replied:
>Doesn't the `*` mean that `" "` may occur an arbitrary number of times?

Perhaps, when I read the RFC yesterday I understood it could happen only once. I'll skip all whitespaces

RonEld commented on library/x509_create.c:
```diff
@@ -86,62 +92,141 @@ static const char *x509_at_oid_from_name( const char *name, size_t name_len )
+                        goto exit;
+                    }
+                    ret = dn_store_named_data( head, oid, data, d, &s,
+                                               &dn_cur_state, &c, end );
```
I agree it's readable, but if we got to this point that we change the state, the the condition is either `c` is 'c' or we have reached the end. In this case, we will fail in the check:
~~~
if( *c == '\\' && c != end )
~~~

hanno-arm replied:
@RonEld Agreed, but as a matter of principle, would you mind changing it? Alternatively, we might add a `break` after the state change.

RonEld replied:
Yes, I'll change it.

RonEld commented on library/x509_create.c:
```diff
@@ -86,62 +92,141 @@ static const char *x509_at_oid_from_name( const char *name, size_t name_len )
 int mbedtls_x509_string_to_names( mbedtls_asn1_named_data **head, const char *name )
 {
     int ret = 0;
     const char *s = name, *c = s;
```
Honestly I am not sure what `s` and `d` stand for...
I will try to understand from the code

RonEld commented:
@hanno-arm I believe I have addressed all your comments, and I also added some additional tests that found new issues(fixed them)

hanno-arm commented on library/x509_create.c:
```diff
@@ -86,62 +92,161 @@ static const char *x509_at_oid_from_name( const char *name, size_t name_len )
+    /*
+     * s points to the current string name.
+     * c points to the current character in the string being parsed.
+     * d points to the end of the current parsed named data.
```
This too could be expanded. Suggestion: `d acts as a write pointer into the temporary 'data' buffer which holds the unescaped and unquoted DN values found in 'name'.`

RonEld replied:
But it points to the end of the data, so I wouldn't call it a write pointer

hanno-arm replied:
@RonEld It is a write pointer, isn't it? It starts with `data` and is increased with every character you copy. There's no static pointer to the end of `data` - instead, it's checked at every copy that `d - data` hasn't yet exceeded the size of `data`.

hanno-arm commented on library/x509_create.c:
```diff
@@ -86,62 +92,161 @@ static const char *x509_at_oid_from_name( const char *name, size_t name_len )
     int ret = 0;
+    /*
+     * s points to the current string name.
+     * c points to the current character in the string being parsed.
```
Can we rename `c` to `cur` to reflect this meaning?

RonEld replied:
ok

hanno-arm commented on library/x509_create.c:
```diff
@@ -86,62 +92,161 @@ static const char *x509_at_oid_from_name( const char *name, size_t name_len )
 {
     int ret = 0;
+    /*
+     * s points to the current string name.
```
I don't find this very helpful yet. Suggestion: rename `s` to `start` and document `'start' is a pointer into the 'name' buffer. Depending on the parsing state, it either points to the beginning of the currently parsed DN OID tag or the currently parsed DN value.`

RonEld replied:
ok, but it is reasonable that it stood for string.

hanno-arm commented on library/x509_create.c:
```diff
@@ -86,62 +92,161 @@ static const char *x509_at_oid_from_name( const char *name, size_t name_len )
+                    }
+                    s = c + 1;
+                    d = data;
+                    if( *s == '\"' )
```
This doesn't handle white spaces correctly: We should skip whitespaces before and after the equality sign. With the present code, something like `A= "B"` would be treated as an unquoted value.

RonEld replied:
I'll add a test case for this as well

hanno-arm commented on library/x509_create.c:
```diff
@@ -86,62 +92,161 @@ static const char *x509_at_oid_from_name( const char *name, size_t name_len )
         {
-            if( ( oid = x509_at_oid_from_name( s, c - s ) ) == NULL )
+            /*
+             *Currently parsing a DN tag.
```
Minor: Space missing after "*".

hanno-arm commented on library/x509_create.c:
```diff
@@ -86,62 +92,161 @@ static const char *x509_at_oid_from_name( const char *name, size_t name_len )
+                        ret = MBEDTLS_ERR_X509_INVALID_NAME;
+                        goto exit;
+                    }
+                    ret = dn_store_named_data( head, oid, data,
```
The whitespace skipping is not handled both here and in `dn_store_named_data`. Given that it is also used before the `=` which separates OID and value, I think there should be a uniform whitespace-skipping function which is used in all states, and `dn_store_named_data` should be removed in favor of a direct call to `mbedtls_asn1_store_named_data`.

RonEld replied:
But then we will have some duplication in every state.

hanno-arm replied:
@RonEld Why? We could have a static function `skip_spaces` or similar which is called whenever potential whitespace needs to be skipped.

hanno-arm replied:
`dn_stored_named_data` is essentially a white space skip + call to `mbedtls_asn1_store_named_data`.

RonEld replied:
It also checks that `d - data` doesn't exceed the size limit.
Personally, I thought it would be more readablehaving it all ion the same function, but if you object, I'll remove it

hanno-arm replied:
I object to having it non-uniform (and hence less readable) which it currently is. 

My preference would be to remove `dn_store_named_data()` and potentially introduce a white space skipping function instead. It doesn't matter that it's only a few lines of code - the compiler may inline it if it thinks it's worth it - but it'll nonetheless increase readability.

RonEld commented:
@hanno-arm I believe I have addressed your issues.
I have moved some code to be in their specific state, although I don't think it makes the code any clearer.
I have also fixed some pre-existing issues related to white spaces.

Note that this PR is starting to exceed its original purpose, which was to fix handling commas.
I don't mind refactoring and handling other existing issues, I just think it requires tickets to be raised and to be handled in other PRs

hanno-arm commented on library/x509_create.c:
```diff
@@ -86,73 +92,208 @@ static const char *x509_at_oid_from_name( const char *name, size_t name_len )
-                goto exit;
+                if( *cur == '=' )
+                {
+                    if( ( oid = x509_at_oid_from_name( start, tag_len ) ) == NULL )
```
As mentioned above, we're still not handling whitespaces prior to `=` correctly, e.g. `A =B` wouldn't be parsed correctly as far as I see.

hanno-arm replied:
@RonEld Sorry, I missed that. A whitespace terminates the current tag, though, while the present code keeps parsing; for example, `A BC=D` would be parsed as the OID `A B` currently.

RonEld replied:
I have also added a [test case](https://github.com/ARMmbed/mbedtls/pull/784/files#diff-0807f152469bce0d302ec7282849e58bR131) for this

RonEld replied:
I mean, whitespaces in the middle of a tag is not permitted

hanno-arm replied:
@RonEld Yes, but we're currently not checking for that, or do we?

RonEld replied:
`x509_at_oid_from_name()` will return an error in this case, as returned in the test"X509 String to Names #19"

hanno-arm replied:
I think the current code wouldn't handle ` A=B` correctly, because it'd parse the OID as ` ` instead of `A`.

RonEld replied:
you may have a point. I'll add a test case for this and fix if needed

RonEld commented:
@hanno-arm I fixed the preceding white spaces issue. Please review

hanno-arm commented on library/x509_create.c:
```diff
@@ -86,73 +92,214 @@ static const char *x509_at_oid_from_name( const char *name, size_t name_len )
+        cur++;
+    }
+
+    if( dn_cur_state != dn_in_tag )
```
With this, we could accept a DN string like `A=B C`, ignoring the `C`. We might fix this by requiring `start == end` I think.

RonEld replied:
don't think it will work. `start` will always be equal to `end + 1` when we finish the operation.
I think that the whole whitespace handling requires refactoring. How about adding 4 more states:
* `dn_before_tag'  - will be initialization state, will allow whitespace, and transition to `dn_in_tag` once characters arrive. otherwise - error.
*  `dn_after_tag` - will be transitioned from `dn_in_tag` if there is white space. Allows only whitespace and `=`
* `dn_before_value` - same as `dn_before_tag` but for value.
* `dn_after_value` - same as `dn_after_tag` but for value.

hanno-arm replied:
@RonEld I fully agree that a proper solution to this needs enhancing the state machine, and I'd be happy if you'd take the effort to implement it. In your proposal, I think `dn_after_value` needs no separation from `dn_before_tag`, or does it?

RonEld replied:
`dn_after_value` should allow whitespaces, and transition to `dn_before_tag` once a comma arives.
`dn_before_tag` shouold allow whitespaces and transition to `dn_in_tag` once valid characters arrive.

I'll add the changes

RonEld commented:
@hanno-arm I believe I have resolved all the issues related to whitespaces.

sbutcher-arm commented:
@RonEld - Looks like you have conflicts and need some rework.

RonEld commented:
I rebased to resolve conflicts.
I believe I carefully integrated the changes from https://github.com/ARMmbed/mbedtls/pull/1641 in the conflict handling.
@hanno-arm  please review

sbutcher-arm commented:
@AndrzejKurek / @andresag01 / @hanno-arm - Can you all please re-review? Thanks.

andresag01 commented on library/x509_create.c:
```diff
@@ -125,81 +136,352 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+                    /*
+                     * Fill the current data pointer with the unescaped character.
+                     */
+                    *(d++) = *cur;
```
Style: Please add spaces before and after ) (

andresag01 commented on library/x509_create.c:
```diff
@@ -125,81 +136,352 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+                    cur++;
+                    start = cur + 1;
+                    dn_cur_state = dn_after_quoted_value;
+                    continue;
```
Cant we just leave it to break out of here normally and the cur will be incremented automatically?

RonEld replied:
I don't think so. As you can see, we are skipping the quotation mark and modify `start` to point to `+2` from original `cur

andresag01 commented on library/x509_create.c:
```diff
@@ -125,81 +136,352 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+                    cur++;
+                }
+
+                if( *cur == ',' || cur == end )
```
Same comment about the 1 byte overread. You might want to reverse the operands to || to fix this here.

andresag01 commented on library/x509_create.c:
```diff
@@ -125,81 +136,352 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+                    dn_cur_state = dn_after_unquoted_value;
+                    continue;
+                }
+                else if( *cur == ',' || cur == end )
```
Same comment about the 1 byte overread

andresag01 commented on library/x509_create.c:
```diff
@@ -125,81 +136,352 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+                    /*
+                     * Fill the current data pointer with the unescaped character.
+                     */
+                    *(d++) = *cur;
```
Style: Please add spaces before and after brackets

andresag01 commented on library/x509_create.c:
```diff
@@ -125,81 +136,352 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+                 */
+                while( cur < end && *cur == ' ' )
+                {
+                    *(d++) = *cur;
```
Style: please add spaces before and after brackets.

andresag01 commented on library/x509_create.c:
```diff
@@ -125,81 +136,352 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+                    cur++;
+                }
+
+                if( *cur == ',' || cur == end )
```
same comment about the 1 byte overread

andresag01 commented on tests/data_files/Makefile:
```diff
@@ -844,6 +852,7 @@ server1.v1.der: server1.v1.crt
 	$(OPENSSL) x509 -inform PEM -in $< -outform DER -out $@
 all_final += server1.v1.crt server1.v1.der
 
+
```
Why is this newline here?

RonEld replied:
probably by mistake

RonEld commented on library/x509_create.c:
```diff
@@ -39,6 +39,17 @@ typedef struct {
+    dn_in_quoted_value = 5,
+    dn_after_quoted_value = 6,
+    dn_after_unquoted_value = 7
+} x509_dn_parse_state;
```
I know, but I think it's more readable this way

RonEld commented:
@andresag01 Thank you for the review!
Yes, this began as a fix for a minor issue, but I was requested to rework the function to make it conform the RFC more. I will try to make helper functions

RonEld commented on library/x509_create.c:
```diff
@@ -86,73 +97,345 @@ static const char *x509_at_oid_from_name( const char *name, size_t name_len )
-
-            /* Check for valid escaped characters */
-            if( c == end || *c != ',' )
+            break;
```
If you mean literraly `A=B`, then `x509_attr_descr_from_name()` should fail, befcause `A` is not a know OID.

If you mean because no spaces, then this is valid

RonEld commented on library/x509_create.c:
```diff
@@ -125,81 +136,352 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
 
+static int x509_dn_check_special_chars( const char **cur, const char *end )
+{
+    const char *c = *cur;
```
To make it easier when comparing the contents with the special character. At the moment the special character is only `,`, but this function will be enhanced for all the other special characters allowed in distinguished names.

Without having this local variable, we will need to do something like `*(*cur) !=  ','` which I think is more confusing for debugging than using a single pointer.

RonEld commented on library/x509_create.c:
```diff
@@ -125,81 +136,352 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+                    start++;
+                }
+
+                if( *cur == '=' )
```
I am not sure we have an overread, because in worst case, `cur == end`. In that case, since the input is a string, then end is the terminating null character

AndrzejKurek commented:
@RonEld Please re-request reviews once you're satisfied with your work.

RonEld commented:
@andresag01 @hanno-arm I have refactored `mbedtls_x509_string_to_names()` to make it more readable.
Please review

andresag01 commented on library/x509_create.c:
```diff
@@ -125,81 +125,285 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+    /*
+     * We have reached here in the following conditions:
+     * 1. *cur == end
+     * 2. in_quot == 1 && *( *cur ) == '"'
```
Is this really the case? There is a check just before failing if `in_quot == 1`

RonEld replied:
Yes, perhaps the description is not clear. 
When we are `in_quot` and reached a closing `"~ character, we set `in_quot` to false, and exit the parser loop. This is the second condition ( even if we already set `in_quot` to false.)
If at the end of the loop, `in_quot` is true, then we haven't got a closing `"` character, and thus fail

andresag01 replied:
Thanks for the explanation, I agree with your assessment. However, I also think that the comment is very misleading, because if I am not familiar with the code and read this comment, I will probably get confused with the condition immediately above.

To make things clearer, how about if instead of just writing the conditions we also write an english version of what we expect? For example, instead of just `*cur == end` we have `The end of the input buffer was reached (*cur == end)`.

andresag01 commented on library/x509_create.c:
```diff
@@ -125,81 +125,285 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
-        {
-            *(d++) = *c;
+    /*
+     * Current position must be wither end or separator character.
```
Typo: wither

RonEld commented on library/x509_create.c:
```diff
@@ -125,81 +125,285 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+    /*
+     * Skip whitespaces
+     */
+    while( *cur < end && *( *cur ) == ' ' )
```
your <CR> character is not seen:)

Perhaps, but I don't think this is part of the scope of this PR... This PR is a bug fix for handling commas ( wnd whitespaces along the way).
We will do an enhancement for supporting all other special characters, and missing characters, which I think is more suitable

andresag01 replied:
Happy with that. Better to fix the immediate problem now and add more features later if we need/want them

RonEld commented:
@andresag01 I have fixed the typos, and addressed your questions.
Please review

andresag01 commented:
@RonEld: Thanks for reworking and answering my questions. I just left a minor comment.

RonEld commented:
@andresag01 I have updated the comment  per your request.

Not that there is a conflict in the data files Makefile. 
In addition, this PR requires some history cleanup. Do you want me to clean the history before your review or after? (Same question regarding the conflict resolution)

andresag01 commented:
@RonEld: Thanks for modifying the comment.

>In addition, this PR requires some history cleanup. Do you want me to clean the history before your review or after? (Same question regarding the conflict resolution)

Personally, I would prefer if this is done before the next round of reviews. Otherwise we will have to review it multiple times even though there were only history and rebase changes.

RonEld commented:
@andresag01 Sure, since you are the reviewer, I'll do what you prefer.
I just think that if a history change is done before you review all the changes ( e.g. https://github.com/ARMmbed/mbedtls/pull/784/commits/16127f367868f65f9e18cb5a7d8264505d423d54), then it will be more difficult for you to review only the changes after the rebase

RonEld commented:
@andresag01 I have rebased to resolve conflicts, and cleared history.
Once approved, I will update the backports

andresag01 commented on library/x509_create.c:
```diff
@@ -125,81 +125,288 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+    /*
+     * Read the next characters to evaluate as a valid tag.
+     */
+    while( *cur < end && *( *cur ) != '=' && *( *cur ) !=' ' )
```
Minor: Please add a space between `=` and `'`

andresag01 commented on library/x509_create.c:
```diff
@@ -125,81 +125,288 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+
+    if( *cur == end )
+    {
+        ret = MBEDTLS_ERR_X509_INVALID_NAME ;
```
Minor: please remove space between `E` and `;`

andresag01 commented on library/x509_create.c:
```diff
@@ -125,81 +125,288 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+        /*
+         * Fill the current data pointer with the unescaped character.
+         */
+        *( d++ ) = *( *cur );
```
Question: Do we need to check that we have not ran out of space in the buffer pointed to by `d` before writing it?

RonEld replied:
I don't think so. Because we first fill `d` and then we increment it by one. After that, we check that if `d - data == MBEDTLS_X509_MAX_DN_NAME_SIZE` we return error.
So, there is already a check that it isn't increased.

But, perhaps there is a special corner case where there is a valid string, followed only by white spaces, exceeding MBEDTLS_X509_MAX_DN_NAME_SIZE. In this case, value_len will be correct, but we will still write to `d` the whitespace. 
So, a followup question is:
In this corner case, do we want to return an error, or do we just don't write anymore to d, and use the correct value indicated by `value_len`?
I favor the second option

andresag01 commented on library/x509_create.c:
```diff
@@ -125,81 +125,288 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+    /*
+     * The only valid character now is '='
+     */
+    if( *cur == end || *( *cur ) != '=' )
```
Question: I have been wondering about this. From the RFC and the comments you have through the code, it looks that a stand-alone attribute is a valid DN. For example, something like "C, D, Z=hello" should be acceptable. Right? However, it looks to me that this is not actually handled by the code. Is this also the case?

Also, i do not think we have test cases for this.

Note: I dont mind if what I pointed out above is not supported, but it would be great if this was documented somewhere in the doxygen documentation of the function (and possibly elsewhere in the code)

(P.S. I do not think this was handled in the previous code either.)

RonEld replied:
But this is not a valid DN, AFAIK. There is always a key and a value

andresag01 replied:
mmm, Ok perhaps I am missing something. I was just trying to compose the productions in the grammar at RFC 1779 to parse a single tag. It looks to me like this can be done, but I am not sure its correct or that they are actually used in practice. For example, parsing the DN `atag,btag" could be done using:

```
<name> -> <named-component> ( <spaced-separator> )
<named-component> -> <attribute>
<attribute> -> <string>
<string> -> *( <stringchar> | <pair> )
etc...
```

RonEld replied:
As for tests, there is [this one](https://github.com/ARMmbed/mbedtls/pull/784/files#diff-0807f152469bce0d302ec7282849e58bR156), but perhaps we could add another tests, with the seperator character after

RonEld replied:
Perhaps I am missing something, but the RFC states:
>The keyword specification is optional in the BNF, but mandatory for
   this specification.  This is so that the same BNF may be used for the
   related specification on User Friendly Naming [5].  When this
   specification is followed, the attribute type keywords must always be
   present.

And ( from the BNF grammar)
~~~
<attribute> ::= <string>
           | <key> <optional-space> "=" <optional-space> <string>
~~~

andresag01 replied:
Ok

andresag01 commented on library/x509_create.c:
```diff
@@ -125,81 +125,288 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
-            cur->val.tag = attr_descr->default_tag;
+    if( value_len == 0 )
+    {
+        ret = MBEDTLS_ERR_X509_INVALID_NAME ;
```
Style: Please remove space before `;`

andresag01 commented on library/x509_create.c:
```diff
@@ -125,81 +125,288 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+     */
+    if( in_quot == 1 )
+    {
+        ret = MBEDTLS_ERR_X509_INVALID_NAME ;
```
Style: Please remove space before `;`

andresag01 commented on library/x509_create.c:
```diff
@@ -125,81 +125,288 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+    }
+
+    if( ( *attr_descr = x509_attr_descr_from_name
+                        ( start, *cur - start ) ) == NULL )
```
Minor suggestion: In this cases where the assignment and conditional do not fit in the same line, I would suggest simply doing the assignment in a separate statement to help readability. For example:

```
*attr_descr = x509_attr_descr_from_name( start, *cur - start );
if(  *attr_descr == NULL )
```

RonEld commented:
@andresag01 I have addressed your comments, and added some more tests.
I haven't squashed them yet, for clearer review. Do you what me to squash them to previous commits prior to you review?

RonEld commented on library/x509_create.c:
```diff
@@ -125,81 +125,296 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+        /*
+         * Fill the current data pointer with the unescaped character.
+         */
+        if( d - data < MBEDTLS_X509_MAX_DN_NAME_SIZE )
```
I discussed this [here](https://github.com/ARMmbed/mbedtls/pull/784#discussion_r265135002)

The idea is, if all the extra characters are only trailing whitespaces, which we ignore, then this is a valid input. Once there is a single character which is not a whitespace, then we return an error

RonEld commented on library/x509_create.c:
```diff
@@ -125,81 +125,296 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+         * value length until the character is  not a whitespace
+         */
+        if( *( *cur ) != ' ' )
+            value_len = d - data;
```
yes. These are trailing whitespaces which we ignore. If there are additional whitespaces, then we don't ignore them

RonEld commented on library/x509_create.c:
```diff
@@ -125,81 +125,296 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+        if( *( *cur ) != ' ' )
+            value_len = d - data;
+
+        if( value_len >= MBEDTLS_X509_MAX_DN_NAME_SIZE )
```
I'm sorry, but I don't understand what you mean

andresag01 replied:
I think in the first paragraph, I was describing a situation that would cause a failure even though there is enough space in the data buffer to hold the input. (Note that A and B are simple characters). Also, it would be nice to have a test case for this, which I do not think we currently have.

I think in the second paragraph I am pointing out that the condition `value_len > MBEDTLS_X509_MAX_DN_NAME_SIZE` will never be true. (Note that the == is not there on purpose) It is dead code because `value_len = d - data` in line 321, but `d - data` can never be greater than `MBEDTLS_X509_MAX_DN_NAME_SIZE` because of the check inline 313.

In summary, I am suggesting that we should rearrange a bit the code here to make sure that these strange corner cases do not happen and also to make the code easier to read. For example, replacing the lines:
```c
        /*
         * Fill the current data pointer with the unescaped character.
         */
        if( d - data < MBEDTLS_X509_MAX_DN_NAME_SIZE )
            *( d++ ) = *( *cur );
        /*
         * A whitespace can be part of the string,
         * but it can also be a trailing whitespace, so don't set the
         * value length until the character is  not a whitespace
         */
        if( *( *cur ) != ' ' )
            value_len = d - data;

        if( value_len >= MBEDTLS_X509_MAX_DN_NAME_SIZE )
        {
            ret = MBEDTLS_ERR_X509_INVALID_NAME;
            goto exit;
        }
```
with something like:
```c
        /*
         * Fill the current data pointer with the unescaped character.
         */
        if( d - data == MBEDTLS_X509_MAX_DN_NAME_SIZE && *( *cur ) != ' ' )
        {
            /* The data buffer is not large enough to hold the value */
            ret = MBEDTLS_ERR_X509_INVALID_NAME;
            goto exit;
        }
        else if( d - data < MBEDTLS_X509_MAX_DN_NAME_SIZE )
        {
            /*
             * Assume that the current character is part of the value tag, but
             * this might actually not be the case if it is just trailing whitespace
             */
            *( d++ ) = *( *cur );
        }

        /*
         * A whitespace can be part of the string,
         * but it can also be a trailing whitespace, so don't set the
         * value length until the character is  not a whitespace
         */        
        if ( *( *cur ) != ' ' )
            value_len = d - data;
```

RonEld replied:
But the first paragraph is actually a failure, because  in this case length of `A<lots of whitespaces>B` is `1 (A) + BEDTLS_X509_MAX_DN_NAME_SIZE -1 ( whitespaces) + 1 ( B) = BEDTLS_X509_MAX_DN_NAME_SIZE + 1`
There is test case for this [here](https://github.com/ARMmbed/mbedtls/pull/784/files#diff-0807f152469bce0d302ec7282849e58bR170)

As for the second paragraph, I believe I meant the following:
` d - data == MBEDTLS_X509_MAX_DN_NAME_SIZE - 1` -> `d++` -> ` d - data == MBEDTLS_X509_MAX_DN_NAME_SIZE`
After that, if current character is not white space, we now have `value_len == MBEDTLS_X509_MAX_DN_NAME_SIZE` which should cause an error I believe.

Anyway, I like your suggestion better, as it's clearer

RonEld replied:
There is minor issue with your suggestion:
`sizeof(data) == MBEDTLS_X509_MAX_DN_NAME_SIZE ` but the longest input string can be 255, to have the terminating null character.
I'll change your suggestion a bit

RonEld commented on library/x509_create.c:
```diff
@@ -125,81 +125,296 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+        if( *( *cur ) != ' ' )
+            value_len = d - data;
+
+        if( value_len >= MBEDTLS_X509_MAX_DN_NAME_SIZE )
```
yes : https://github.com/ARMmbed/mbedtls/pull/784/files#diff-0807f152469bce0d302ec7282849e58bR167

RonEld commented on library/x509_create.c:
```diff
@@ -125,81 +125,296 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
-                mbedtls_asn1_store_named_data( head, oid, strlen( oid ),
-                                               (unsigned char *) data,
-                                               d - data );
+            ret = MBEDTLS_ERR_X509_INVALID_NAME;
```
This is actually same behavior as before

andresag01 replied:
Alright, lets leave it like that then as we do not want an API change.

RonEld commented on library/x509_create.c:
```diff
@@ -125,81 +125,296 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+        /*
+         * A whitespace can be part of the string,
+         * but it can also be a trailing whitespace, so don't set the
+         * value length until the character is  not a whitespace
```
ok

RonEld commented on library/x509_create.c:
```diff
@@ -125,81 +125,296 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+        /*
+         * A valid tag can only be followed by a '=' or whitespace.
+         */
+        if( *( *cur ) == ',' )
```
as you said, `x509_attr_descr_from_name()` will return an error in such a case. I don't think we should take spacial attention for this, in opposed to regular characters that do not make a valid attribute

andresag01 replied:
Ok

andresag01 replied:
Ok

RonEld commented on tests/suites/test_suite_x509write.data:
```diff
@@ -87,19 +87,85 @@ depends_on:MBEDTLS_SHA1_C:MBEDTLS_RSA_C:MBEDTLS_PKCS1_V15:MBEDTLS_DES_C:MBEDTLS_
+X509 String to Names #27 (name with long trailing whitespaces)
+mbedtls_x509_string_to_names:"C=NL                                                                                                                                                                                                                                    ":"C=NL":0
+
+X509 String to Names #27 (name with many whitespaces in the middle)
```
Minor: should change to `#28`

andresag01 replied:
Personally, I would prefer to remove the numbering altogether. I think its unecessary and its very easy for them to get out of sync. However, they are already there, so lets keep them

RonEld replied:
me too, but I just continued the original numbering

andresag01 commented on tests/suites/test_suite_x509write.data:
```diff
@@ -87,19 +87,85 @@ depends_on:MBEDTLS_SHA1_C:MBEDTLS_RSA_C:MBEDTLS_PKCS1_V15:MBEDTLS_DES_C:MBEDTLS_
+mbedtls_x509_string_to_names:"C=NL,ABC=BAD, OU=PolarSSL":"":MBEDTLS_ERR_X509_UNKNOWN_OID
+
+X509 String to Names #27 (name with long trailing whitespaces)
+mbedtls_x509_string_to_names:"C=NL                                                                                                                                                                                                                                    ":"C=NL":0
```
@RonEld: You pointed out that this test covers the case case of very long strings. However, I think the data buffer is of length MBEDTLS_X509_MAX_DN_NAME_SIZE which is hard-coded to 256 by default. But this test only has a DN of length 232 so I do not think the buffer will actually be overflowed. Perhaps I missed something?

I think it would also be good to have a test has a value of length exactly 256 and the character at position 255 is a non whitespace. This is just a corner case and it would be good to make sure we do not miss any data

RonEld replied:
my mistake in counting whitespaces

RonEld commented on library/x509_create.c:
```diff
@@ -116,16 +117,21 @@ int mbedtls_x509_string_to_names( mbedtls_asn1_named_data **head, const char *na
-
             /* Check for valid escaped characters */
-            if( c == end || *c != ',' )
+            if( ( c+1 ) == end || *( c+1 ) != ',' )
```
AS for your first comment, yes, there are other special characters that needs to be addressed. But these should be addressed in a different PR IMHO, as supporting these would be considered an enhancement, while this is still considered a bug fix.

RonEld commented:
@andresag01 I addressed your comments, but I'll squash some commits to clear history

andresag01 commented on library/x509_create.c:
```diff
@@ -310,21 +310,27 @@ static int x509_dn_parse_value( const char **cur, const char *end,
+            ret = MBEDTLS_ERR_X509_INVALID_NAME;
+            goto exit;
+        }
+        else if( d - data < MBEDTLS_X509_MAX_DN_NAME_SIZE -1 )
```
Minor: Please add space before and after `-`

RonEld commented:
@andresag01 Thank you for your review
I fixed the style issue you pointed out, I cleared history, and rebased to resolve conflicts.

Please review, as your approval was dismissed

daverodgman commented:
This fix is covered by #5861  - closing
